### PR TITLE
Upgrade Android SDK to 34 "UpsideDownCake"

### PR DESCRIPTION
**Pull request by gmackall**
~**This should not land until https://github.com/flutter/buildroot/pull/790 (re)lands, and I swap the buildroot url back to the latest commit.**~ ~Reland of PR to update buildroot at https://github.com/flutter/buildroot/pull/792. ~ <- landed, and changed the buildroot commit to the latest in DEPS

Upgrades to android api 34

Also:
1. Upgrades to java 17 in DEPS/ci, because the linter now requires it.
2. Stops running some roboelectric tests on android apis 16-18, because Roboelectric indicated those versions were unsupported and we don't support them either.
3. Applies the four trivial new suggestions from the newer linter.
4. Updates the baseline lint to include the new non-trivial lint ([fixed in a different PR](https://github.com/flutter/engine/pull/47817/files)).
5. Changes some instances where we were hardcoding android apis as numbers (e.g. `sdk = 16') to use version codes (see the [comment below](https://github.com/flutter/engine/pull/47609#issuecomment-1800308658)).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: DEPS [hunk 1]
```diff
@@ -262,7 +262,7 @@ allowed_hosts = [
 ]
 
 deps = {
-  'src': 'https://github.com/flutter/buildroot.git' + '@' + '5ced4367fee2721f6eda34802dc28b1335637c63',
+  'src': 'https://github.com/gmackall/buildroot.git' + '@' + '302769cc2f56a0898ce67274ac8543c8b80e5d9a',
```
Comment (gmackall) [inline DEPS:265]: This is to test https://github.com/flutter/buildroot/pull/790 before landing

File: tools/android_lint/bin/main.dart [hunk 2]
```diff
@@ -73,7 +73,7 @@ Future<int> runLint(ArgParser argParser, ArgResults argResults) async {
 <!-- WILL AUTOMATICALLY FIND ALL .java FILES AND INCLUDE THEM HERE       -->
 <project>
   <sdk dir="${androidSdkDir.path}" />
-  <module name="FlutterEngine" android="true" library="true" compile-sdk-version="android-T">
+  <module name="FlutterEngine" android="true" library="true" compile-sdk-version="android-U">
```
Comment (gmackall) [inline tools/android_lint/bin/main.dart:76]: `android-U` is a guess here. I'm assuming the format is just first letter of the version name, but I can't find documentation

**Comment by gmackall** [comment]
Some tests are failing with the message
```
ninja: Entering directory `/b/s/w/ir/cache/builder/src/out/android_profile'
ninja: error: '../../third_party/android_tools/ndk/sources/android/cpufeatures/cpu-features.c', needed by 'obj/third_party/android_tools/ndk/sources/android/cpufeatures/cpu_features.cpu-features.o', missing and no known rule to make it
```
That file exists for me locally, and the build succeeds. Need to figure out whats going on here, guessing I need to dive into ci/recipes to figure out

**Comment by gmackall** [comment]
> Some tests are failing with the message
> 
> ```
> ninja: Entering directory `/b/s/w/ir/cache/builder/src/out/android_profile'
> ninja: error: '../../third_party/android_tools/ndk/sources/android/cpufeatures/cpu-features.c', needed by 'obj/third_party/android_tools/ndk/sources/android/cpufeatures/cpu_features.cpu-features.o', missing and no known rule to make it
> ```
> 
> That file exists for me locally, and the build succeeds. Need to figure out whats going on here, guessing I need to dive into ci/recipes to figure out

The answer here was that the cipd package creation script doesn't work on m1 macs, as long as we specify an ndk version <23. A new version (34v3) should fix the problem

**Comment by gmackall** [comment]
Many roboelectric tests were failing because they couldn't be run on jelly bean api levels, ex:

```
io.flutter.plugin.platform.SingleViewPresentationTest > returnsOuterContextInputMethodManager[17] FAILED
    java.lang.RuntimeException: java.lang.RuntimeException: Failed to parse package ../../../../../out/android_jit_release_x86/robolectric_tests/build/intermediates/apk_for_local_test/debugUnitTest/apk-for-local-test.ap_: ../../../../../out/android_jit_release_x86/robolectric_tests/build/intermediates/apk_for_local_test/debugUnitTest/apk-for-local-test.ap_ (at Binary XML file line #5): Requires newer sdk version #19 (current version is #17)
        at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:68)
        at org.robolectric.android.internal.AndroidTestEnvironment.loadAppPackage_measured(AndroidTestEnvironment.java:410)
        at org.robolectric.android.internal.AndroidTestEnvironment.lambda$loadAppPackage$3(AndroidTestEnvironment.java:382)
        at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
        at org.robolectric.android.internal.AndroidTestEnvironment.loadAppPackage(AndroidTestEnvironment.java:382)
        at org.robolectric.android.internal.AndroidTestEnvironment.installAndCreateApplication(AndroidTestEnvironment.java:279)
        at org.robolectric.android.internal.AndroidTestEnvironment.lambda$createApplicationSupplier$0(AndroidTestEnvironment.java:251)
        at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
        at org.robolectric.android.internal.AndroidTestEnvironment.lambda$createApplicationSupplier$1(AndroidTestEnvironment.java:248)
        at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:183)
        at org.robolectric.RuntimeEnvironment.getApplication(RuntimeEnvironment.java:79)
        at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:216)
        at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:340)
        at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:281)
        at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:99)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

        Caused by:
        java.lang.RuntimeException: Failed to parse package ../../../../../out/android_jit_release_x86/robolectric_tests/build/intermediates/apk_for_local_test/debugUnitTest/apk-for-local-test.ap_: ../../../../../out/android_jit_release_x86/robolectric_tests/build/intermediates/apk_for_local_test/debugUnitTest/apk-for-local-test.ap_ (at Binary XML file line #5): Requires newer sdk version #19 (current version is #17)
            at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:61)
            ... 18 more
```

We don't support these api levels, so I've changed the tests to stop testing on those api levels. Let me know if you are opposed.

I also changed some places where we were using numbers to represent an api level (e.g. `sdk = 16`) to use build codes (e.g. `sdk = Build.VERSION_CODES.KITKAT`). I'm planning on preferring this format whenever I encounter a place to choose, unless people have other opinions.

**Comment by gmackall** [comment]
Looks like linting requires java 17
```
Preparing project.xml...
Wrote project.xml, starting lint...
Using JAVA_HOME=../third_party/java/openjdk
Error: Unable to initialize main class com.android.tools.lint.Main
Caused by: java.lang.UnsupportedClassVersionError: com/android/tools/lint/client/api/IssueRegistry has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
```

File: shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java [hunk 3]
```diff
@@ -517,8 +518,14 @@ private CharSequence getClipboardData(PlatformChannel.ClipboardContentFormat for
...
         ClipData.Item item = clip.getItemAt(0);
+        AssetFileDescriptor assetFileDescriptor;
         if (item.getUri() != null)
-          activity.getContentResolver().openTypedAssetFileDescriptor(item.getUri(), "text/*", null);
+          assetFileDescriptor =
+              activity
+                  .getContentResolver()
+                  .openTypedAssetFileDescriptor(item.getUri(), "text/*", null);
+        CharSequence charSequence = item.coerceToText(activity);
+        if (assetFileDescriptor != null) assetFileDescriptor.close();
```
Comment (gmackall) [inline shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java:528]: The other linter suggestions were straightforward to resolve, but I'm not sure if this is the right fix for this one.
The linter error was 
```
../../../flutter/shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java:521: Error: This AssetFileDescriptor should be freed up after use with #close() [Recycle]
          activity.getContentResolver().openTypedAssetFileDescriptor(item.getUri(), "text/*", null);
```
I can't actually tell what the AssetFileDescriptor is doing here, unless its creation somehow interacts with [`ClipData.Item.coerceToText`](https://developer.android.com/reference/android/content/ClipData.Item#coerceToText(android.content.Context))? I just decided to close it after getting the `CharSequence`, but I'm not sure if thats right.

Comment (gmackall) [reply shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java]: I'm moving this to a new PR

**Review by reidbaker — APPROVED** [review]

**Pull request merged**